torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1167,60 +856,335 @@ def _get_ignored_params(
+        this method, the module must be either fully on the CPU or fully on a
+        non-CPU device.
+        """
+        devices = set(
+            param.device for param in self._get_orig_params(module, ignored_params)
         )
+        if len(devices) > 1:
+            raise RuntimeError(
+                f"FSDP only supports single device modules but got params on {devices}"
+            )
+
+    def _get_compute_device(
+        self,
+        module: nn.Module,
+        ignored_params: Set[nn.Parameter],
+        device_id: Optional[Union[int, torch.device]],
+    ) -> torch.device:
+        """
+        Determines and returns this FSDP instance's compute device. If the
+        module is already on a non-CPU device, then the compute device is that
+        non-CPU device. If the module is on CPU, then the compute device is the
+        current device.
+
+        The ``device_id`` argument should be what the user passed into the FSDP
+        constructor to produce a faithful error. If ``device_id`` is not
+        ``None``, then we enforce that it must either be (1) the same as the
+        compute device, (2) the same index as the compute device's index, or
+        (3) the general unindexed device. Hence, ``device_id`` is simply a way
+        for the user to have FSDP move the module to device, but that target
+        device is fixed to the current device.
+
+        NOTE: For now, the compute device is always a CUDA GPU device with its
+        explicit index.
+
+        Precondition: ``_check_single_device_module()``.
+        """
+        if not torch.cuda.is_available():
+            raise RuntimeError("FSDP does not support CPU only execution")
+        # If the module is on GPU already, then that GPU device has priority
+        # over the current device
+        param = next(self._get_orig_params(module, ignored_params), None)
+        if param is not None and param.device.type == "cuda":
+            compute_device = param.device
+        else:
+            compute_device = torch.device("cuda", torch.cuda.current_device())
+        if device_id is not None:
+            device = (
+                device_id
+                if isinstance(device_id, torch.device)
+                else torch.device(device_id)
+            )
+            if device == torch.device("cuda"):
+                warnings.warn(
+                    f"FSDP got the argument `device_id` {device_id} on rank "
+                    f"{self.rank}, which does not have an explicit index. "
+                    f"FSDP will use the current device {torch.cuda.current_device()}. "
+                    "If this is incorrect, please explicitly call `torch.cuda.set_device()` "
+                    "before FSDP initialization or pass in the explicit device "
+                    "index as the `device_id` argument."
+                )
+            elif device != compute_device:
+                raise AssertionError(
```

Comment:
curious, is there any reason for using `RuntimeError` for CPU only module and `AssertionError` here?

Reply:
This used to be an `assert`, so I preserved that originally. I changed this to a `RuntimeError` now to be consistent with the other error. It might even be better to make this a `ValueError` since it means the specified `device_id` was a bad value.